src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.ForEachAsync.cs

```diff
@@ -271,7 +268,7 @@ private static Task ForEachAsync<TSource>(IAsyncEnumerable<TSource> source, int
                         }
                         finally
                         {
-                            state.Lock.Release();
+                            state.ReleaseLock();
                         }
 
```

Comment:
Here the `state.Cancellation.IsCancellationRequested` could be checked once again, to reduce the cancellation latency in this scenario:
1. The cancellation occurred while the `state.Enumerator.MoveNextAsync` was in flight.
2. The enumerator doesn't observe the token.
3. The `body` doesn't observe the token either.

Reply:
I don't think that's necessary.  If neither the enumerator nor the body pays any attention to cancellation, the consumer isn't showing a lot of care for that use case and we don't need to be super aggressive about adding such checks.  If they want it and their body doesn't do it, they can just include that check in the body.

Reply:
I am not thinking about people who are calling non-cancellable APIs inside the `body`. I am thinking about people who don't even know why the `body` lambda has a second argument, and what is the purpose of this argument. I've seen quite a lot of people totally ignoring this argument in StackOverflow questions and answers.

Reply:
If they're ignoring the CancellationToken argument, they have bigger problems than a single up-front polling check in that case.

Reply:
Indeed. Btw [this answer](https://stackoverflow.com/questions/15136542/parallel-foreach-with-asynchronous-lambda/68901782#68901782) is probably the most upvoted answer on StackOverflow, demonstrating the `Parallel.ForEachAsync` API. The `token` is ignored. At least the parameter is named correctly, instead of `_`. :-)